### ISSUE TITLE
Safari: Bindings not working sometimes

### ISSUE THREAD

**Issue by dnish** [ManuelDeLeon/viewmodel#188]
Sorry for opening the next issue, but I've just seen, that the bindings aren't working 100% on Safari. This is my template:

    <input type="text" class="form-control animated" placeholder="{{_ "type_link"}}" id="linkInput" {{b "change:hasError(false), value:linkInput, class:{is-error:hasError, rubberBand:hasError}"}}/>

The linked VM:

    Template.mytemp.viewmodel({

    linkInput: null,
    hasError: false,

    });

On about 70% of my page calls with Safari, linkInput is null if I type in a link. There is no error message in the console. This only happens in Safari, on Chrome and FF everything is working fine.

**Comment by ManuelDeLeon** [comment]
Make a repro without the placeholder, change, class, rubberBand, etc.

**Comment by dnish** [comment]
I'll do that tomorrow, I just found out, that it has something to do with the package <code>martino:iron-router-i18n</code>. I'm using it to deliver my Meteor app in different languages. This is what I do:

    Template.layout.viewmodel({

    onRendered() {

        //Language things

        var normalUserAgent = I18NConf.getNavigatorLanguage();


        if (normalUserAgent.indexOf('-') > 0) {
            normalUserAgent = normalUserAgent.substring(0, normalUserAgent.indexOf('-'));
        }


        if (!Router.getLangCode()) {
            I18NConf.setLanguage(normalUserAgent);
        }
    }
});

This will do the following: If a german user visits "localhost:3000" he will get redirected to "localhost:3000/de" after rendering the layout. If I deactivate the code, VM works fine on Safari, after reactivating it, about 70% of all page calls create a VM that stopped working.

**Comment by dnish** [comment]
Mhhh...it works fine when I remove all TAPi18n helpers from my template, so for example this is my original template:

    <p id="selectedFormat" {{b "click:changeFormat"}}>{{_ "type"}}: <u>{{formatFormat}}</u></p>
                    <div class="input-group"><input type="text" class="form-control animated" placeholder="{{_ "type_link"}}" id="linkInput" {{b "change:hasError(false), value:linkInput, class:{is-error:hasError, rubberBand:hasError}"}}>
                        <div class="input-group-btn">
                            <button type="button" class="btn btn-default" id="convertButton" {{b "click:addConvert"}}>{{_ "convert"}}</button>
                        </div>
                    </div>

If I change it to

    <p id="selectedFormat" {{b "click:changeFormat"}}>File type: <u>{{formatFormat}}</u></p>
                    <div class="input-group"><input type="text" class="form-control animated" placeholder="Just type in your link..." id="linkInput" {{b "change:hasError(false), value:linkInput, class:{is-error:hasError, rubberBand:hasError}"}}>
                        <div class="input-group-btn">
                            <button type="button" class="btn btn-default" id="convertButton" {{b "click:addConvert"}}>Start conversion</button>
                        </div>
                    </div>

On every "failed page call" you can see that the site is in English for a seconds, then switches to German. After that, my VM bindings aren't working anymore.

**Comment by ManuelDeLeon** [comment]
Did this used to work? If so then make a repro please.

**Comment by dnish** [comment]
Hey, you can use the repro above https://github.com/dnish/viewmodel-repro

You just need to remove the <code>placeholder="{{_ "placeholder"}}"</code> in hello template. Then, everything should work again on Safari.

**Comment by ManuelDeLeon** [comment]
I don't get it. What's that underscore?

**Comment by dnish** [comment]
It's the tap18n helper - https://github.com/TAPevents/tap-i18n#the-tap-i18n-helpers

It gives me the translation for the specific key in my *.i18n.json's.

**Comment by daveeel** [comment]
Just curious. Is that even a valid syntax ?
```html
placeholder="{{_ "placeholder"}}"
```
Did you try
```html
placeholder="{{_ 'placeholder'}}"
```

**Comment by fvpDev** [comment]
+1 @daveeel

**Comment by dnish** [comment]
@daveeel Why should it be invalid syntax? In pure HTML this would be invalid, but a SpaceBar helper should work. In my case, everything works fine except in Safari most of the time. I think WebStorm would show me if this is invalid.

**Comment by ManuelDeLeon** [comment]
I have a problem with this one because I don't have a mac and the safari for windows doesn't reproduce the problem. I'm open to suggestions.

**Comment by dnish** [comment]
@ManuelDeLeon I think the tapi18n helper causes the problem. When I get the failed binding, my page shows me for a second the english translation, not the german. After this one second, it switches to the german translation but the binding stops working.

**Comment by daveeel** [comment]
@dnish , yes, you are right, it's not a syntax issue.

I tried your repro and can confirm the issue does exist in Safari. 

On visiting `http:localhost:3000/de`, with the place holder code, VM binding has no effect:
@ManuelDeLeon , here's the element tree after I manually `$('input').attr('value','234')` in console:
```<input placeholder="Etwas eintippen" type="text" b-id="6" value="234">```
Even after that. `ViewModel.findOne('hello').myText('hello')` has no effect.

Not sure how VM works internally. But seems something breaks ...

**Comment by ManuelDeLeon** [comment]
I'd be happy to work on this one if someone gives me a way to reproduce it on Windows or Linux. I've heard folk tales of Mac VMs but I've never seen one that works.

**Comment by daveeel** [comment]
Found this Mac in cloud. Not sure if it worth your effort to try: https://www.macincloud.com/features/tools/tools

I searched the doc and found no placeholder binding, will that be worth adding?

**Comment by dnish** [comment]
Well, the placeholder was just an example. It also doesn't work on other attributes when we use TAPi18n within VMs template.

**Comment by ManuelDeLeon** [comment]
@daveeel Thanks for the link. I think it's worth a try. 

> I searched the doc and found no placeholder binding, will that be worth adding?

You can change any attribute with the `attr` binding. That includes `placeholder`.

Now that you mention it, I think there should be a way for you to add "attributes as bindings". Something like `ViewModel.addAttributeBinding(['src','href','readonly','placeholder'])`.

I'll think about this one.

@dnish I'll got on a trip tomorrow so I don't know when/if I'll have to work on this in the next 2 weeks.

**Comment by dnish** [comment]
@ManuelDeLeon Just having the issue now on my Windows 10 machine running Firefox 44, you can check it here: http://www.convert.as/de/vimeo - if you click on the blue "Umwandeln" button, nothing happens...

**Comment by ManuelDeLeon** [comment]
Okay, I have absolutely no clue what Safari and TAPi18n are doing. Here's the situation:

ViewModel creates a custom Blaze helper called `b` (that's what you normally use with `{{b "value: name"}}`. The `b` helper basically returns an id and adds a hook to `Blaze.currentView.onViewReady` which uses the id to find the element on the page (and then bind to it).

For whatever reason Safari and TAPi18n mess it up. When you access a `/de` route, the `b` helper returns the id of the element and sets the callback for `onViewReady`. So far so good, but then the element gets re-rendered (generating a new id for it). By the time the `onViewReady` callback runs, it can't find the element on the page (there's a new one). And here's where it gets weird, the new element does set a callback for `Blaze.currentView.onViewReady` so in theory it should bind to the new element, but it doesn't. The callback never runs.

I don't think I'll spend more time on this because as far as I can tell ViewModel isn't doing anything wrong.

In case anyone is interested in investigating check out the `getBindHelper` method in the `viewmodel.coffee` file, specifically on line 183 and put breakpoints/console.logs showing the `bindId` variable before `Blaze.currentView.onViewReady ->` and inside the callback.

I'll reopen this ticket if someone can figure out why the element's `{{b ` custom helper runs again when the element also has TAPi18n's `{{_ ` custom helper AND the `/de` route is used AND it runs on Safari. Bonus: why the `onViewReady` callback isn't called for the second time the `b` helper is invoked.

**Comment by dnish** [comment]
Thank you for you reply @ManuelDeLeon. I don't know if this helps you, but after adding <code>mrt:facebook-sdk</code> and doing this in my ViewModel:

````
Template.layout.viewmodel({

onRendered() { 
    window.fbAsyncInit = function() {
            FB.init({
                appId      : '1078872282155354',
                status     : true,
                xfbml      : true
            });
        };
}
});

````

the issue also happens in Google Chrome all the time.

**Comment by ManuelDeLeon** [comment]
Does that mean bindings break even without i18n?

**Comment by dnish** [comment]
I will check that in a repro, I just added it to my current i18n project.

**Comment by ManuelDeLeon** [comment]
btw, try adding `fbAsyncInit` in a js file outside a template.

**Comment by dnish** [comment]
@ManuelDeLeon Seems to be related to i18n, I can't reproduce it in a new project with <code>mrt:facebook-sdk</code> and <code>manuel:viewmodel</code>. Works fine on every reload.

**Comment by dnish** [comment]
Okay, this issue now happens also in Google Chrome and is definitely caused by i18n, even without the facebook package I get this error.

I've tried the following scenarios:

1. Calling "myapp.com/de" directly: ViewModel fails
2. Calling "myapp.com" and getting redirected to "myapp.com/de": Everything works fine
3. Changing browser language to "EN" and calling "myapp.com": Everything works fine

I will look deeper into that i18n-iron-router package if I got time, there must be "that difference" that causes this issue.
